ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala

```diff
@@ -0,0 +1,145 @@
+
+package com.digitalasset.platform.sandbox.stores
+
+import java.time.Instant
+
+import com.daml.ledger.participant.state.v1.AbsoluteContractInst
+import com.digitalasset.daml.lf.data.Ref.{Party, TransactionIdString}
+import com.digitalasset.daml.lf.data.Relation.Relation
+import com.digitalasset.daml.lf.transaction.Node.GlobalKey
+import com.digitalasset.daml.lf.transaction.GenTransaction
+import com.digitalasset.daml.lf.value.Value
+import com.digitalasset.daml.lf.value.Value.AbsoluteContractId
+import com.digitalasset.ledger.WorkflowId
+import com.digitalasset.ledger.api.domain.PartyDetails
+import com.digitalasset.platform.sandbox.stores.ActiveLedgerState._
+import com.digitalasset.platform.sandbox.stores.ledger.SequencingError
+import scalaz.syntax.std.map._
+
+case class InMemoryActiveLedgerState(
+    activeContracts: Map[AbsoluteContractId, ActiveContract],
+    divulgedContracts: Map[AbsoluteContractId, DivulgedContract],
+    keys: Map[GlobalKey, AbsoluteContractId],
+    parties: Map[Party, PartyDetails])
+    extends ActiveLedgerState[InMemoryActiveLedgerState] {
+
+  override def lookupActiveContract(cid: AbsoluteContractId): Option[ActiveContract] =
+    activeContracts.get(cid)
+
+  override def lookupContract(cid: AbsoluteContractId): Option[Contract] =
+    activeContracts.get(cid).orElse[Contract](divulgedContracts.get(cid))
+
+  override def keyExists(key: GlobalKey) = keys.contains(key)
+
+  /**
+    * Updates divulgence information on the given active contract with information
+    * from the already existing divulged contract.
+    */
+  private def copyDivulgences(ac: ActiveContract, dc: DivulgedContract): ActiveContract =
+    ac.copy(divulgences = ac.divulgences.unionWith(dc.divulgences)((l, _) => l))
+
+  override def addContract(
+      c: ActiveContract,
+      keyO: Option[GlobalKey]): InMemoryActiveLedgerState = {
+    val newKeys = keyO match {
+      case None => keys
+      case Some(key) => keys + (key -> c.id)
+    }
+    divulgedContracts.get(c.id) match {
+      case None =>
+        copy(
+          activeContracts = activeContracts + (c.id -> c),
+          keys = newKeys
+        )
+      case Some(dc) =>
+        copy(
+          activeContracts = activeContracts + (c.id -> copyDivulgences(c, dc)),
+          divulgedContracts = divulgedContracts - c.id,
+          keys = newKeys
+        )
+    }
+  }
+
+  override def removeContract(cid: AbsoluteContractId, keyO: Option[GlobalKey]) = keyO match {
+    case None => copy(activeContracts = activeContracts - cid)
+    case Some(key) => copy(activeContracts = activeContracts - cid, keys = keys - key)
+  }
+
+  override def addParties(newParties: Set[Party]): InMemoryActiveLedgerState =
+    copy(parties = newParties.map(p => p -> PartyDetails(p, None, true)).toMap ++ parties)
+
+  override def divulgeAlreadyCommittedContract(
+      transactionId: TransactionIdString,
+      global: Relation[AbsoluteContractId, Party],
```

Comment:
Please rename `global` (there is no `local`?) to `divulgences` or something in that direction.